### PR TITLE
Add more information about extensible structures

### PR DESCRIPTION
**Pull request by david-a-wheeler**
Add information from discussions into comments. The text is mostly by
Mario Carneiro and @Benjub. See:
https://github.com/metamath/set.mm/issues/1314

Signed-off-by: David A. Wheeler <dwheeler@dwheeler.com>

### CODE REVIEW

**Comment by benjub** [comment]
Thanks @david-a-wheeler for documenting this. While I see nothing false in your proposal, I would like to modify it a bit but I won't have the time to do it before Christmas. Also, @digama0 is probably the best person for such a summary, if he has time.  I don't know if you can wait, or if you want to merge soon and that after, I propose a new PR ? @nmegill ?
In any case:
* exensible --> extensible
* the hypothetical case of NNmnd has an actual analogue in set.mm, which is preferable to use in these explanations: this is df-zring (by the way: the underlying set of NNmnd is NN0 in the set.mm notation).
* I would remove the reference to manifolds, or at least put it in a more general remark that, roughly, extensible structures work well for "sets with extra structure" (roughly: the objects of concrete categories), but have limitations for more general structures (for instance, for manifolds, where |\`s would not work). However, extensible structures are sufficient for the structures that set.mm currently considers, and offer a good compromise for a goal-oriented formalization.

File: set.mm [hunk 1]
```diff
@@ -179333,6 +179333,73 @@ other components such as a multiplicative operation for rings (and still
...
+  Should we come up with some new structure in the future that we
+  wish ` CC ` to inherit, then we change the definition of ` CCfld ` ,
+  reprove all the slot extraction theorems, add a new one,
+  and that's it. None of the other downstream theorems have to change.
+
+  The restriction operation ` |`s ` is not very useful for manifolds.
+  Manifold embeddings or immersions are something else entirely; you
+  clearly need a function to perform that mapping so the implicit subset
+  morphism doesn't cut it. In that case, the typical solution would be to
+  have a set ( M ManifoldImmersion N ) for encoding all these details.
```
Comment (digama0) [inline set.mm]: This paragraph doesn't make much sense out of context. I suggest dropping it.

File: set.mm [hunk 2]
```diff
@@ -179333,6 +179333,73 @@ other components such as a multiplicative operation for rings (and still
...
+  X is a Y via a bunch of forgetful functors) can
+  cause serious performance problems for automated tooling,
+  and the resulting proofs would be painful to look at
+  directly (in the case of Lean, they are long past the level where
+  people would find it acceptable to look at them directly).
+  Metamath is working under much stricter conditions than this,
+  and it has still managed to achieve about the same level of flexibility.
+
+  To create a substructure of a given exensible structure, you can simply
+  use the multifunction restriction operator for extensible structures ` |`s `
```
Comment (tirix) [inline set.mm]: Warning, you have to double the ` within math formula embedded in a comment!
(same in other instances below)

**Comment by david-a-wheeler** [comment]
@digana0 - I'm happy to drop it for now, but I'd really like to document more rationale in general. Can you suggest other text instead?

**Comment by david-a-wheeler** [comment]
Thanks for your comments all!  I tried to respond to all comments, and I expanded the description further.

Extensible structures are used all over the place, but the way they're done in set.mm is not something you'll find in a math textbook. Thus, while I think documenting is good everywhere, I think it's especially important that extensible structures be explained (both what they are and the rationale for the approach).

File: set.mm (hunk 2 above)
Comment (david-a-wheeler) [reply set.mm]: Fixed!

File: set.mm (hunk 1 above)
Comment (david-a-wheeler) [reply set.mm]: Dropped.

**Comment by david-a-wheeler** [comment]
(Travis is *really* unusually slow to get going right now, looks like they are working off a backlog.)

**Comment by david-a-wheeler** [comment]
Everything passes, I think this is ready for now. We can always add further improvements later.

**Comment by benjub** [comment]
Thanks @david-a-wheeler for these explanations.  Everything is correct, but I think there are a few things where the emphasis is not put at the right place (for instance the passage about Moore collections). I still think @digama0 is the best person to edit these paragraphs if he has time. Maybe I'll try to propose something in a subsequent PR, but not in the coming days.

**Comment by david-a-wheeler** [comment]
@benjub - if anyone has further improvements, I greatly welcome them. Just create a new pull request with those changes.  I certainly agree that Mario could add much more. But I figured that some improvement was better than none.

**Comment by benjub** [comment]
By the way, @david-a-wheeler, since you wrote about Moore collections: they could be used in your study of binary relations (I know that for the moment, it's only to have the definition written, but maybe you'll develop it), see the last paragraph of my message https://github.com/metamath/set.mm/pull/1286#issuecomment-560756122.  This is a typical process: you first prove that a collection is stable under intersections, and then you can take the "closure", or the "group/ring/etc. generated by" as the intersection of all the sets containing the given set: it is guaranteed to be of the same sort, by stability under intersections, and to be the smallest one. Same thing for closed subsets, convex subsets, sigma-algebras, etc.

**Pull request merged**